lib/permission-api.js

```diff
@@ -0,0 +1,274 @@
+////////////////////////////////////////////////////////////////////////////
+//
+// Copyright 2017 Realm Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+////////////////////////////////////////////////////////////////////////////
+
+'use strict';
+
+const url_parse = require('url-parse');
+const managementSchema = require('./management-schema');
+
+function generateUniqueId() {
+  const uuid = 'xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx'.replace(/[xy]/g, function (c) {
+    var r = Math.random() * 16 | 0, v = c == 'x' ? r : (r & 0x3 | 0x8);
+    return v.toString(16);
+  });
+  return uuid;
+}
+
+const permissionSchema = [
+  {
+    name: 'Permission',
+    properties: {
+        id:            { type: 'string' },
+        updatedAt:     { type: 'date' },
+        userId:        { type: 'string' },
+        path:          { type: 'string' },
+        mayRead:       { type: 'bool' },
+        mayWrite:      { type: 'bool' },
+        mayManage:     { type: 'bool' },
+    }
+  }
+];
+
+// Symbols are not supported on RN yet, so we use this for now:
+const specialPurposeRealmsKey = '_specialPurposeRealms';
+
+function getSpecialPurposeRealm(user, realmName, schema) {
+  if (!user.hasOwnProperty(specialPurposeRealmsKey)) {
+    user[specialPurposeRealmsKey] = {};
+  }
+
+  if (user[specialPurposeRealmsKey].hasOwnProperty(realmName)) {
+    return Promise.resolve(user[specialPurposeRealmsKey][realmName]);
+  }
+
+  const url = url_parse(user.server);
+  if (url.protocol === 'http:') {
+    url.set('protocol', 'realm:');
+  } else if (url.protocol === 'https:') {
+    url.set('protocol', 'realms:');
+  } else {
+    throw new Error(`Unexpected user auth url: ${user.server}`);
+  }
+
+  url.set('pathname', `/~/${realmName}`);
+
+  const config = {
+    schema: schema,
+    sync: {
+      user,
+      url: url.href
+    }
+  };
+
+  const _Realm = user.constructor._realmConstructor;
+
+  return new Promise((resolve, reject) => {
+    _Realm._waitForDownload(config, (error) => {
```

Comment:
why not use Realm.open. Seems this is duplicating the same logic

Reply:
I guess I could, it's just that it hasn't been added at this point, so I would have to reference the function directly through extensions.js somehow. Seemed hacky anyway.

Reply:
got it. 10x